plugins/inputs/logfile/tailersrc_test.go

```diff
@@ -296,6 +305,65 @@ func TestOffsetDoneCallBack(t *testing.T) {
 	}
 }
 
+func TestTailerSrcFiltersSingleLineLogs(t *testing.T) {
+	original := multilineWaitPeriod
+	defer resetState(original)
+	done := make(chan struct{})
+	var consumed int32
+	file, statefile := setupTailer(t, nil, done, &consumed, defaultMaxEventSize)
+	defer os.Remove(file.Name())
+	defer os.Remove(statefile.Name())
+
+	n := 100
+	matchedLog := "ERROR: this has an error in it."
+	unmatchedLog := "Some other log message"
+	publishLogsToFile(file, matchedLog, unmatchedLog, n, 0)
+
+	// Removal of log file should stop tailersrc
+	if err := os.Remove(file.Name()); err != nil {
+		t.Errorf("failed to remove log file '%v': %v", file.Name(), err)
+	}
+	<-done
+	assertExpectedLogsPublished(t, n, int(consumed))
+}
+
+func TestTailerSrcFiltersMultiLineLogs(t *testing.T) {
+	original := multilineWaitPeriod
+	defer resetState(original)
+	done := make(chan struct{})
+	var consumed int32
+	file, statefile := setupTailer(
+		t,
+		regexp.MustCompile(`\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}[Z+\-]\d{2}:\d{2}`).MatchString,
+		done,
+		&consumed,
+		defaultMaxEventSize,
+	)
+	defer os.Remove(file.Name())
+	defer os.Remove(statefile.Name())
+
+	n := 20
+	// create log messages ahead of time to save compute time
+	buf := bytes.Buffer{}
+	buf.WriteString("This has a matching log in the middle of it")
+	buf.WriteString(strings.Repeat("\nfoo", 2))
+	buf.WriteString("\nHere is the ERROR that should be matched")
+	buf.WriteString(strings.Repeat("\nfoo", 2))
+	matchedLog := buf.String()
+	buf.Reset()
+
+	unmatchedLog := "This should not be matched." + strings.Repeat("\nbar", 5)
+
+	publishLogsToFile(file, matchedLog, unmatchedLog, n, 100)
+
+	// Removal of log file should stop tailersrc
+	if err := os.Remove(file.Name()); err != nil {
+		t.Errorf("failed to remove log file '%v': %v", file.Name(), err)
+	}
+	<-done
+	assertExpectedLogsPublished(t, n, int(consumed))
+}
+
```

Comment:
By the way, about your comment "Not sure why, but truncation only occurs for multiline logs."  I believe that's because the buffered reader itself is [declared with a max line size](https://github.com/aws/amazon-cloudwatch-agent/blob/abdd1653cdcdf2cba7a91fa9615ca8deb683c7b2/plugins/inputs/logfile/tail/tail.go#L475-L475).  So if you have a single line that is longer than the max size, it will be split into multiple log events anyway.  (But then I would expect you would see more than `n` log events in `consumed`.)

Reply:
That makes sense and explains the weird numbers I saw when I was writing a test for truncated logs. Wasn't aware of that. Noted.

Reply:
Did we decide against having a truncated logs test at all then?  I think it would be good to verify the logic that filters are applied on the truncated (multi-line) message.  (You can choose a different max event size so you don't have to allocate 256KB of memory to run the test.)

Reply:
That was the conclusion from yesterday. I did tweak the max event size to be smaller, but it still fails consistently in CI, but works without issue locally. Also, the flakiness doesn't seem related to the sleep timing like the multiline log test.

Reply:
Argh, that's annoying.  OK I will approve then.